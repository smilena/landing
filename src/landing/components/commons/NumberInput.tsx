import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { setUserData, UserData } from '../../../redux-state/UserDataSlice';
import { setContinueToNavigation } from "../../../redux-state/NavigationSlice";
import { useEffect } from "react";
import { RootState } from '../../../store';

export type NumberInputProps = {
  id: string;
  label: string;
  max?: number;
}

const Components = {
  Wrapper: styled.div`
    display: flex;
    flex-direction: column;
  `,
  Label: styled.label`
    color: #504437;
    font-size: 18px;
  `,
  Input: styled.input`
    color: #504437;
    font-size: 18px;
    height: 30px;
    border: 1px solid #504437;
    border-radius: 3px;
  `,
}

export const NumberInput = ({id, label, max = 50}: NumberInputProps) => {
  
  const dispatch = useDispatch();
  const userData:UserData = useSelector((state:RootState) => state.userData);
  const [value, setValue] = useState(userData[id  as keyof UserData ]);

  useEffect(() => {
    dispatch(setContinueToNavigation(value > 0));
  }, [value]);

  const onChangeHandler = (e: React.FormEvent<EventTarget>) => {
    const { value }  = e.target as HTMLInputElement;
    setValue(Number(value));
    dispatch(setUserData({id, value}));
  }
  
  return (
    <Components.Wrapper>
      <Components.Label htmlFor={id}>{label}</Components.Label>
      <Components.Input
        data-testid="input-number"
        type="number" 
        id={id} 
        max={max}
        value={Number(value)}
        onChange={onChangeHandler}
      />
    </Components.Wrapper>
  )
}
