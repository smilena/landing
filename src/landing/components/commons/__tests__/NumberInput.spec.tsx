import { render, cleanup, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { NumberInput } from '../NumberInput';

const mockStore = createMockStore([]);

afterEach(cleanup);

describe('Given NumberInput component', () => {
  
  test('should render correctly with initial state', () => {
    const state = {
      userData: {
        floor: 6,
      }
    };
    const store = mockStore(state);
    const { container } = render(
      <Provider store={store}>
        <NumberInput id={'floor'} label={'Número de piso'}/>
      </Provider>);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId('input-number').getAttribute('value')).toEqual('6');
    expect(screen.getByText('Número de piso')).toBeInTheDocument();
  });

});
