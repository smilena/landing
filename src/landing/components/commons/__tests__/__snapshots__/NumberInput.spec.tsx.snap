// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given NumberInput component should render correctly with initial state 1`] = `
<div>
  <div
    class="sc-bczRLJ eaUlQc"
  >
    <label
      class="sc-gsnTZi bfxakb"
      for="floor"
    >
      NÃºmero de piso
    </label>
    <input
      class="sc-dkzDqf dWwYGf"
      data-testid="input-number"
      id="floor"
      max="50"
      type="number"
      value="6"
    />
  </div>
</div>
`;
